### PR TITLE
Refactored TopHeavyHitters and added comments. Fixed small bug.

### PR DESCRIPTION
**Pull request by drewhk**
I fully reviewed TopHeavyHitters again, recorded my notes as comments. Apart from a minor bug it looks completely correct.

### CODE REVIEW

File: akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala [hunk 1]
```diff
@@ -198,35 +310,34 @@ private[remote] final class TopHeavyHitters[T >: Null](val max: Int)(implicit cl
...
   }
 
   /**
    * Remove value from hash-table based on position.
-   *
-   * @param index position to remove
    */
   private def removeHash(index: Int): Unit = {
-    if (index > 0) {
+    if (index >= 0) {
```
Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala]: bug 1

File: akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala [hunk 2]
```diff
@@ -248,39 +359,31 @@ private[remote] final class TopHeavyHitters[T >: Null](val max: Int)(implicit cl
...
+  /**
+   * Weight of lowest heavy hitter, if a new inserted item has a weight greater than this, it is a heavy hitter.
+   * This gets the index of the lowest heavy hitter from the top of the heap (lowestHitterIndex) if there is any
+   * and looks up its weight.
+   * If there is no entry in the table (lowestHitterIndex returns a negative index) this returns a zero weight.
+   */
   def lowestHitterWeight: Long = {
     val index: Int = lowestHitterIndex
-    if (index > 0) {
+    if (index >= 0) {
```
Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala]: bug 2

**Comment by drewhk** [comment]
cc @bambuchaAdm

**Comment by drewhk** [comment]
This might explain duplicate entries in a compression table (gaps) reported in https://github.com/akka/akka/issues/21780 (https://jenkins.akka.io:8498/job/akka-nightly/3851) because removeHash might not have removed an entry with hash mapping to zero. Not completely sure though, but I added an extra test that fails on the original code.

**Comment by drewhk** [comment]
No, it cannot explain those gaps. These bugs can cause:
 - remaining garbage in hashmap, which can lead in worst case to infinite loop in `insert`, but the entry is properly removed from the heap so will not enter the snapshot
 - reporting wrong lowestHitterWeight, but this can only lead to suboptimal table allocation by evicting someone wrongly.

Also, those gaps are hard to explain given that Map(heavyHitters.snapshot.filterNot(_ == null).zipWithIndex: _*) generates the table. I don't get how that could introduce a gap.

File: akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala [hunk 3]
```diff
@@ -66,44 +131,77 @@ private[remote] final class TopHeavyHitters[T >: Null](val max: Int)(implicit cl
...
+   */
   private def findItemIdx(searchFromIndex: Int, hashCode: HashCodeVal, o: T): Int = {
     @tailrec def loop(index: Int, start: Int, hashCodeVal: HashCodeVal, o: T): Int = {
+      // Scanned the whole table, returned to the start => element not in table
       if (index == start) -1
-      else if (hashCodeVal.get == hashes(index)) {
+      else if (hashCodeVal.get == hashes(index)) { // First check on hashcode to avoid costly equality
         val item: T = items(index)
-        if (item == o) {
+        if (Objects.equals(item, o)) {
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala]: what is the advantage of Objects.equals?

**Review by patriknw — APPROVED** [review]
LGTM, but note that the missing entry in the log that I looked at was value 4 (not 0)

File: akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala (hunk 3 above)
Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala]: It was used everywhere else, I just made it consistent.

**Comment by drewhk** [comment]
@patriknw that is of no relevance since I am talking about the hash index which has nothing to do with the advertisement index.

**Comment by patriknw** [comment]
@drewhk The gap must be because the returned `Array[T]` of `heavyHitters.snapshot` contains a duplicate, which is then removed when turning it into a `Map`.

So two elements in `heap` is somehow pointing to the same slot in `items`, or `items` itself contains duplicates.

**Comment by drewhk** [comment]
> The gap must be because the returned Array[T] of heavyHitters.snapshot contains a duplicate, which is then removed when turning it into a Map.

That is true.

> So two elements in heap is somehow pointing to the same slot in items

I looked the whole code and I have no evidence for such error so far.

> or items itself contains duplicates.

That can happen due to the bug I fixed, but even then, heap would not contain a duplicate entry. 

What I can imagine that when updating an item again that should have been removed (but did not because it was hashed to 0) then it finds the not removed entry, which then points to a wrong heap slot. But then restoring the heap property on the wrong slot will do nothing so this should not cause any of such issues.

I heavily commented the code and what it does, you can look at it, maybe you find something that I missed, but so far I was not able to find any explanation to duplicate entries in that map.

**Comment by drewhk** [comment]
Looking at the original failure again, I catched this log prefix:
```
[ERROR] [11/08/2016 04:00:49.856] [ArteryMultiNodeSpec-remote-2-akka.actor.default-dispatcher-3] [TaskInvocation]
```
What does TaskInvocation mean there? I have never seen this before. I mean, if we access this datastructure concurrently then the error we see can happen very easily. Why is TaskInvocation logged there? I see it comes from dispatcher...

**Comment by drewhk** [comment]
Graaah:
```
akka.remote.artery.compress.InboundCompression$$anonfun$1.apply$mcV$sp(InboundCompressions.scala:291)
	at akka.actor.Scheduler$$anon$2.run(Scheduler.scala:78)
	at akka.actor.LightArrayRevolverScheduler$$anon$2$$anon$1.run(LightArrayRevolverScheduler.scala:102)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:39)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:415)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
It comes from the Scheduler!!!!
So InboundCompression.scala:291 is this:
```
  val schedulerTask: Option[Cancellable] =
    tableAdvertisementInterval match {
      case d: FiniteDuration ⇒
        Some(system.scheduler.schedule(d, d)(runNextTableAdvertisement)(system.dispatcher))
      case _ ⇒
        None
    }
```
Is this safe? @ktoso ?

**Comment by drewhk** [comment]
Ok, so Codecs:374 calls `hitActorRef` which transitively calls `update()` on this table. Meanwhile, in `runNextTableAdvertisement()` calls `prepareCompressionAdvertisement` (InboundCompressions:459) which takes the snapshot. As far as I know Codec is owned by a GraphStage, so this looks like concurrent access that is not safe.

**Comment by patriknw** [comment]
@drewhk ouch, then it's time to implement https://github.com/akka/akka/issues/21475

File: akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala (hunk 3 above)
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala]: must have been some code moved from java I guess.
It's a null-safe way to `.equals()` in java.

**Comment by ktoso** [comment]
> Ok, so Codecs:374 calls hitActorRef which transitively calls update() on this table. Meanwhile, in runNextTableAdvertisement() calls prepareCompressionAdvertisement (InboundCompressions:459) which takes the snapshot. As far as I know Codec is owned by a GraphStage, so this looks like concurrent access that is not safe.

Yeah, definitely seems so - good analysis, thanks Endre!

**Comment by ktoso** [comment]
LGTM

**Comment by ktoso** [comment]
This would be good to include tomorrow, please rebase?

**Comment by drewhk** [comment]
ok

**Comment by drewhk** [comment]
Updated, waiting on validation.

**Pull request merged**